k2/csrc/array_ops.cu

```diff
@@ -212,45 +205,56 @@ bool ValidateRowSplitsAndIds(Array1<int32_t> &row_splits,
   ContextPtr ctx = GetContext(row_splits, row_ids);
   int32_t num_rows = row_splits.Dim() - 1, num_elems = row_ids.Dim();
   if (num_rows < 0 || (num_rows == 0 && num_elems > 0)) return false;
-  if (num_rows == 0 && num_elems == 0) {
-    return row_splits[0] == 0;
-  }
-  int32_t *row_ids_data = row_ids.Data(), *row_splits_data = row_splits.Data();
+  if (row_splits[0] != 0 || row_ids[0] < 0) return false;
+
+  const int32_t *row_ids_data = row_ids.Data(),
+                *row_splits_data = row_splits.Data();
 
   Array1<int32_t> temp_array;
   if (temp == nullptr || temp->Dim() == 0) {
-    temp_array = Array1<int32_t>(row_ids.Context(), 1);
-    temp = &temp_array;
+    temp_array = Array1<int32_t>(ctx, 1);
+  } else {
+    temp_array = temp->Range(0, 1);
   }
-  int32_t *temp_data = temp_array.Data();
-  // The following isn't totally ideal, it would be better to have a single
-  // kernel and avoid latency.  Later we can fix this.
-  if (!ValidateRowSplits(row_splits, temp)) return false;
+  temp = &temp_array;
+  *temp = 0;
 
+  int32_t *temp_data = temp_array.Data();
   auto lambda_check_row_ids = [=] __host__ __device__(int32_t i) -> void {
-    int32_t this_row = row_ids_data[i];
-    if (this_row < 0 || this_row >= num_rows ||
-        this_row < row_splits_data[this_row] ||
-        this_row >= row_splits_data[this_row + 1])
-      *temp_data = 1;
+    // check row_splits
+    bool invalid_splits =
+        (i < num_rows && row_splits_data[i] > row_splits_data[i + 1]);
+    // check row_ids
+    bool invalid_ids =
+        (i < (num_elems - 1) && row_ids_data[i] > row_ids_data[i + 1]);
+    if (invalid_splits || invalid_ids) *temp_data = 1;
+    // check if row_splits and row_ids agree with each other
+    if (i < num_elems) {
+      int32_t this_row = row_ids_data[i];
+      if (this_row < 0 || this_row >= num_rows ||
+          i < row_splits_data[this_row] || i >= row_splits_data[this_row + 1])
+        *temp_data = 1;
+    }
   };
-  Eval(ctx, num_elems, lambda_check_row_ids);
-  return !((*temp)[0]);
+  Eval(ctx, std::max(num_elems, num_rows), lambda_check_row_ids);
+  return (*temp)[0] == 0;
 }
 
-void RowSplitsToRowIds(Array1<int32_t> &row_splits, Array1<int32_t> &row_ids) {
+void RowSplitsToRowIds(const Array1<int32_t> &row_splits,
+                       Array1<int32_t> &row_ids) {
   ContextPtr c = GetContext(row_splits, row_ids);
   int32_t num_elems = row_ids.Dim(), num_rows = row_splits.Dim() - 1;
-  K2_CHECK(num_rows >= 0);
+  K2_CHECK_GE(num_rows, 0);
```

Comment:
An extra check `K2_CHECK_EQ(num_elems, row_splits[-1])`. // psesudo code
may be also added.

Is it required that `num_rows >= 1`?

Reply:
fixed, return false in that case, thanks

Reply:
not we just require num_rows >=0. Empty case is fine